cpp/src/reductions/scan/scan_inclusive.cu

```diff
@@ -64,6 +65,82 @@ rmm::device_buffer mask_scan(const column_view& input_view,
   return mask;
 }
 
+std::unique_ptr<column> inclusive_rank_scan(aggregation const& agg,
+                                            rmm::cuda_stream_view stream,
+                                            rmm::mr::device_memory_resource* mr)
+{
+  auto const& rank_agg = dynamic_cast<rank_aggregation const&>(agg);
+  auto d_order_by      = table_device_view::create(rank_agg._order_by);
+  auto ranks           = make_fixed_width_column(cudf::data_type{cudf::type_to_id<size_type>()},
+                                       rank_agg._order_by.num_rows(),
+                                       mask_state::ALL_VALID,
+                                       stream,
+                                       mr);
+  auto mutable_ranks   = ranks->mutable_view();
+  if (has_nulls(rank_agg._order_by)) {
+    row_equality_comparator<true> row_comparator(*d_order_by, *d_order_by, true);
+    thrust::tabulate(
+      rmm::exec_policy(stream),
+      mutable_ranks.begin<size_type>(),
+      mutable_ranks.end<size_type>(),
+      [row_comparator] __device__(size_type row_index) {
+        return (row_index == 0 || !row_comparator(row_index, row_index - 1)) ? row_index + 1 : 0;
+      });
+  } else {
+    row_equality_comparator<false> row_comparator(*d_order_by, *d_order_by, true);
+    thrust::tabulate(
+      rmm::exec_policy(stream),
+      mutable_ranks.begin<size_type>(),
+      mutable_ranks.end<size_type>(),
+      [row_comparator] __device__(size_type row_index) {
+        return (row_index == 0 || !row_comparator(row_index, row_index - 1)) ? row_index + 1 : 0;
+      });
+  }
```

Comment:
Similar to above, extract common code into lambda with comparator is its parameter.